nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/PublishPulsarRecord_1_X.java

```diff
@@ -0,0 +1,294 @@
+        relationships.add(REL_FAILURE);
+        RELATIONSHIPS = Collections.unmodifiableSet(relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return RELATIONSHIPS;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+
+        final FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final String topic = context.getProperty(TOPIC).evaluateAttributeExpressions(flowFile).getValue();
+
+        if (StringUtils.isBlank(topic)) {
+            getLogger().error("Invalid topic specified {}", new Object[] {topic});
+            session.transfer(flowFile, REL_FAILURE);
+            return;
+        }
+
+        // Read the contents of the FlowFile into a byte array
+        final byte[] messageContent = new byte[(int) flowFile.getSize()];
+        session.read(flowFile, new InputStreamCallback() {
+            @Override
+            public void process(final InputStream in) throws IOException {
+                StreamUtils.fillBuffer(in, messageContent, true);
+            }
+        });
+
+        // Nothing to do, so skip this Flow file.
+        if (messageContent == null || messageContent.length < 1) {
+            session.transfer(flowFile, REL_SUCCESS);
+            return;
+        }
+
+        final RecordReaderFactory readerFactory = context.getProperty(RECORD_READER)
+                .asControllerService(RecordReaderFactory.class);
+
+        final RecordSetWriterFactory writerFactory = context.getProperty(RECORD_WRITER)
+                    .asControllerService(RecordSetWriterFactory.class);
+
+        final Map<String, String> attributes = flowFile.getAttributes();
+        final AtomicLong messagesSent = new AtomicLong(0L);
+
+        try {
+            final InputStream in = new ByteArrayInputStream(messageContent);
+            final RecordReader reader = readerFactory.createRecordReader(attributes, in, getLogger());
+            final RecordSet recordSet = reader.createRecordSet();
+            final RecordSchema schema = writerFactory.getSchema(attributes, recordSet.getSchema());
+            final Producer producer = getWrappedProducer(topic, context).getProducer();
+
+            if (context.getProperty(ASYNC_ENABLED).isSet() && context.getProperty(ASYNC_ENABLED).asBoolean()) {
+               InFlightMessageMonitor bundle = getInFlightMessages(writerFactory, schema, recordSet);
+               this.sendAsync(producer, session, flowFile, bundle);
+               handleAsync(bundle, session, flowFile, topic);
+           } else {
+               messagesSent.addAndGet(send(producer, writerFactory, schema, recordSet));
+               session.putAttribute(flowFile, MSG_COUNT, messagesSent.get() + "");
```

Comment:
These should be `flowfile = session.putAttribute(...` Otherwise AFAIK it's going to drop the new attribute.